homeassistant/components/tuya/cover.py

```diff
@@ -60,6 +62,8 @@ def __init__(self, tuya, platform):
         """Init tuya cover device."""
         super().__init__(tuya, platform)
         self.entity_id = ENTITY_ID_FORMAT.format(tuya.object_id())
+        self._was_closing = False
+        self._was_opening = False
```

Comment:
Can we move this into the pypi package?

Reply:
Sorry, I don't understand what you mean.

Reply:
Ideally the library returns the state of the cover so we don't have `self._was_closing = False`  and `self._was_opening = False` in Home Assistant.

Maybe `self._tuya.cover_state()`

Maybe an enum with the following states.

TuyaCoverState.CLOSING
TuyaCoverState.CLOSED
TuyaCoverState.OPENING
TuyaCoverState.OPENED

Reply:
Tuya API do not provide closing and opening, so this state here is just deducted based on the command send.
I could also open a PR in TuyaHA, but normally code owner take a long time to manage this...

Reply:
Unfortunately we aren't allowed to fake state in Home Assistant.  Please take this out and move it into the pypi package.  I will prioritize a follow up review when the package owner gets the merge done.

Reply:
Wait, I think that I do not properly explain, because I too had forgotten the correct logic.

Tuya provide 3 states:
- 1=Opening
- 2=Closing
- 3=Stopped

What is missing is closed and open, so I was using the 2 variable to implement this logic: 
- if status != 2 and previous was 2 (_was_closing) -> cover is closed (is_closed return true)
- if status != 1 and previous was 1 (_was_opening) -> cover is open (is_closed return false)
- other case return none (means not fully open nor closed)

When cover is manually stopped, the 2 variable related to previous state are reset.

How do you think this 3 state should differently be managed?

Reply:
If the state isn't polled while the cover is opening or closing will the state be incorrect?

Reply:
It will be None because we will only know that is not moving.

Reply:
I ordered one of these for testing.

https://www.amazon.com/Controller-Anywhere-Compatible-Assistant-Required/dp/B07PRRY5NN

Reply:
Edit: I found one I can get tomorrow https://www.amazon.com/gp/product/B07VWNCR14/ref=ppx_yo_dt_b_asin_title_o00_s00?ie=UTF8&psc=1

Reply:
Garage door opener came in, but the integration doesn't find it.

Reply:
Unfortunately this may happen, not all device are supported by this API.

Reply:
If can help, you can refer to [this](https://github.com/ollo69/ha_tuya_custom/issues/7) discussion.
But if this check is an issue, I can finally remove it and always return `None`  from `is_closed` method, that means that returned status will be `Opening`, `Closing `or `Unknown`

Reply:
At first I was hitting the rate limit.  It turns out that calling `tuya.get_access_token()` appears to reset the rate limit at least for the discover call.

Once I got past that, the garage door device was missing from the endpoint.

Reply:
If the `cover` device is something like this https://www.amazon.com/Essenc-Curtain-Blinds-Shutter-Control/dp/B08LPFCTGV/ref=sr_1_15?dchild=1&keywords=tuya+curtain&qid=1603682860&sr=8-15 the state should be `None` (which will be UNKNOWN) and we only trigger open/close and don't know the state.

It looks like we allowed an optimistic mode https://github.com/home-assistant/core/pull/31407 so something like that should be ok as well.

Reply:
the cover is something like one of this two products: https://www.amazon.com/Rollerhouse-Tubular-Rechargeable-Electric-Curtains/dp/B078XDZD1B/ref=sr_1_3?crid=2ME291PH11QNB&dchild=1&keywords=tuya+curtain+motor&qid=1603684005&sprefix=tuya+curt%2Caps%2C247&sr=8-3 or https://www.ajaxonline.co.uk/product/smart-blind-motor/

I have one, so let me know if there's something I need to test.

Reply:
It looks like https://www.amazon.com/Electric-Control-Drapery-compatiable-Broadlink/dp/B07SWRQ7GX/ref=sr_1_2?dchild=1&keywords=tuya+curtain&qid=1603763564&sr=8-2 
or
https://www.amazon.com/Motorized-Roller-36-38MM-Shutter-Control/dp/B0836J7QT9

could be the device as well so its probably ok as-is.

I'd order the second one for testing but I'm about to be traveling so I wouldn't get it in time.  Given the integration is barely working now due to the rate limits, it seems like something we can adjust later if need be.